File: wotd.js
#!/usr/bin/env node

import puppeteer from "puppeteer";
import fs from "node:fs";
import os from "node:os";

const homedir = os.homedir();
const cache_file_date = `${homedir}/.cache/js_wotd_cache_ts.txt`;
const cache_file_word = `${homedir}/.cache/js_wotd_word.json`;
const today = new Date().toDateString();

const fetchWord = async () => {
  // Start a Puppeteer session with:
  // - a visible browser (`headless: false` - easier to debug because you'll see the browser in action)
  // - no default viewport (`defaultViewport: null` - website page will in full width and height)
  const browser = await puppeteer.launch({
    //headless: false,
    headless: true,
    defaultViewport: null,
  });

  const page = await browser.newPage();

  await page.goto("https://www.merriam-webster.com/word-of-the-day", {
    waitUntil: "domcontentloaded",
  });

  const word_data = await page.evaluate(() => {
    const word = document.querySelector("h2.word-header-txt").textContent;
    const word_partofspeech = document.querySelector(
      ".word-attributes span.main-attr",
    ).textContent;
    const word_syllables = document.querySelector(
      ".word-attributes span.word-syllables",
    ).textContent;
    const word_description = document.querySelectorAll(
      ".wod-definition-container p",
    )[0].textContent;
    const word_usage = document.querySelectorAll(
      ".wod-definition-container p",
    )[1].textContent;

    return {
      word,
      word_partofspeech,
      word_syllables,
      word_description,
      word_usage,
    };
  });
  // console.log(`### getting definition of ${word_data.word}`);
  // const word_description2 = fetch(
  //   `https://api.dictionaryapi.dev/api/v2/entries/en/${word_data[0]}`,
  // ).then((desc) => {
  //   console.log(desc);
  //   word_data["word_description2"] = desc;
  // });
  const getDefinition = async () => {
    const response = await fetch(
      ` https://api.dictionaryapi.dev/api/v2/entries/en/${word_data.word} `,
    );
    const json = await response.json();
    // console.log(json[0].meanings[0].definitions[0].definition);
    return json[0].meanings[0].definitions[0].definition;
  };

  getDefinition().then((word_definition) => {
    // console.log(word_definition);
    word_data.word_definition = word_definition;
    // console.log(word_data.word_definition);
    fs.writeFile(cache_file_word, JSON.stringify(word_data), function (err) {
      if (err) throw err;
      // console.log("### saved todays word to cache");
    });
  });

  await browser.close();
};

function readWord() {
  fs.readFile(cache_file_word, "utf8", (err, cache_word) => {
    if (err) {
      // console.error(err);
      fetchWord();
    }

    // console.log(cache_word);
    var cache_word_parsed = JSON.parse(cache_word);
    //     console.log(`${cache_word_parsed["word"]} (${cache_word_parsed["word_partofspeech"]} • ${cache_word_parsed["word_syllables"]})
    // ${cache_word_parsed["word_description"]}
    // ${cache_word_parsed["word_definition"]}
    // ${cache_word_parsed["word_usage"]}
    // `);
    console.log(`${cache_word_parsed["word"]} (${cache_word_parsed["word_partofspeech"]} • ${cache_word_parsed["word_syllables"]})

${cache_word_parsed["word_definition"]}
`);
  });
}

function main() {
  fs.readFile(cache_file_date, "utf8", (err, cache_date) => {
    if (err) {
      // console.error(err);
      fs.writeFile(cache_file_date, today, function (err) {
        if (err) throw err;
        // console.log("### saved todays date to cache");
      });
    }

    // console.log(cache_date);
    if (cache_date === new Date().toDateString()) {
      // console.log("### we should have this cached");
      readWord();
    } else {
      fetchWord(); //.then(readWord());
    }
  });
}

main();

File: wotd.go
//usr/bin/env go run $0 $@ ; exit
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"

	"github.com/adrg/xdg"
	"github.com/gookit/color"
)

type Config struct {
	Version int
	Name    string
	Tags    []string
}

var wd word_data

type word_data struct {
	Date         string `json:"date"`
	Word         string `json:"word"`
	Definition   string `json:"definition"`
	Description  string `json:"description"`
	PartOfSpeech string `json:"partofspeech"`
	Phonetics    string `json:"phonetics"`
	Usage string `json:"usage"`
}

// i hate how many fucking structs i have to use for this but i dont understand go well enough to fix this
type Phonetic struct {
	Text      string `json:"text"`
	Audio     string `json:"audio"`
	SourceURL string `json:"sourceUrl"`
	License   struct {
		Name string `json:"name"`
		URL  string `json:"url"`
	} `json:"license"`
}

type Definition struct {
	Definition string `json:"definition"`
	Synonyms   []any  `json:"synonyms"`
	Antonyms   []any  `json:"antonyms"`
}

type Meaning struct {
	PartOfSpeech string       `json:"partOfSpeech"`
	Definitions  []Definition `json:"definitions"`
	Synonyms     []string     `json:"synonyms"`
	Antonyms     []string     `json:"antonyms"`
}

type License struct {
	Name string `json:"name"`
	URL  string `json:"url"`
}

type DictionaryApiResp struct {
	Word       string     `json:"word"`
	Phonetic   string     `json:"phonetic"`
	Phonetics  []Phonetic `json:"phonetics"`
	Meanings   []Meaning  `json:"meanings"`
	License    License    `json:"license"`
	SourceUrls []string   `json:"sourceUrls"`
}

type WOTDApiResp struct {
	Date         string `json:"date"`
	Word         string `json:"word"`
	PartOfSpeech string `json:"pos"`
	Phonetics    string `json:"ipa"`
	Definition   string `json:"definition"`
}

// func loadConfig() {
// 	doc, fileerr := os.ReadFile(xdg.CacheHome + "/wotd_word.json")
// 	if fileerr != nil {
// 		fmt.Println("### cfg file not exist, whatever")
// 	}

// 	var cfg Config
// 	tomlerr := toml.Unmarshal([]byte(doc), &cfg)
// 	if tomlerr != nil {
// 		panic(tomlerr)
// 	}
// 	fmt.Println("cfg:", cfg)
// }

func fetchDefinition(word string) {
	client := http.Client{
		Timeout: 5 * time.Second,
	}
	resp, err := client.Get("https://api.dictionaryapi.dev/api/v2/entries/en/" + word)
	if err != nil {
		// handle error
		fmt.Println("error fetching definition:", err)
		return
	}

	// read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		// handle error
		fmt.Println("error reading response body:", err)
		return
	}

	// close the body and check for errors
	if cerr := resp.Body.Close(); cerr != nil {
		fmt.Println("error closing response body:", cerr)
	}

	var respdata []DictionaryApiResp

	jsonerr := json.Unmarshal(body, &respdata)
	if jsonerr != nil {
		fmt.Println("could not fetch definition, likely a temporary DictionaryAPI issue. try again later?")
	}

	// Assuming the API returns a list and you want the first item
	if len(respdata) > 0 {
		data := respdata[0]

		for _, meaning := range data.Meanings {
			for _, definition := range meaning.Definitions {
				// fmt.Println("got definition:", definition.Definition)

				wd.Definition = definition.Definition
				break
			}
		}

	} else {
		fmt.Println("fetchDefinition: no data found")
	}
}

func fetchWord() {
	// get the current date
	timenow := time.Now()

	// format date to YYYY-MM-DD
	date := timenow.Format("2006-01-02")

	// initialize http client
	client := http.Client{
		Timeout: 2500 * time.Millisecond,
	}

	// send a request to WOTD API using today's date
	resp, err := client.Get("https://api.wotd.site/query?date=" + date)
	if err != nil {
		// handle error
		fmt.Println("error fetching word:", err)
		return
	}

	// read response response body into a variable
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		// handle error
		fmt.Println("error reading response body:", err)
		return
	}

	// close the body and check for errors
	if cerr := resp.Body.Close(); cerr != nil {
		fmt.Println("error closing response body:", cerr)
		panic(cerr)
	}

	var respdata WOTDApiResp

	jsonerr := json.Unmarshal(body, &respdata)
	if jsonerr != nil {
		fmt.Println("error while parsing word data, likely a temporary WOTD API issue. try again later?")
		panic(jsonerr)
	}

	wd.Date = respdata.Date
	wd.Word = respdata.Word
	wd.PartOfSpeech = respdata.PartOfSpeech
	wd.Phonetics = respdata.Phonetics
	wd.Definition = respdata.Definition

	// try getting a better definition
	fetchDefinition(wd.Word)

	// convert wd struct to json
	content, marshalErr := json.Marshal(wd)
	if marshalErr != nil {
		fmt.Println(marshalErr.Error())
	}

	// save word data to cache
	wordWriteErr := os.WriteFile(xdg.CacheHome+"/wotd_word.json", content, 0644)
	if wordWriteErr != nil {
		fmt.Println("could not save word to cache")
		fmt.Println(marshalErr.Error())
	}
}

func readWord() {
	content, err := os.ReadFile(xdg.CacheHome + "/wotd_word.json")
	if err != nil {
		fmt.Println("### cached word file not exist, calling fetchWord()")
	}

	jsonerr := json.Unmarshal(content, &wd)
	if jsonerr != nil {
		fmt.Println(err.Error())
	}

	fmt.Printf("%s (%s • %s)\n%s\n%s\n",
		color.Bold.Sprintf(wd.Word),
		color.OpItalic.Sprintf(wd.PartOfSpeech),
		wd.Phonetics, color.Gray.Sprintf(wd.Definition),
		color.OpItalic.Sprintf(color.Gray.Sprintf(wd.Usage)))

}

func cacheIsUpdated() bool {
	// read wotd cache file
	fileContent, fileReadErr := os.ReadFile(xdg.CacheHome + "/wotd_word.json")
	if fileReadErr != nil {
		// fmt.Println("### wotd cache ts file doesnt exist")
		return false
	}

	var cacheFileJSON word_data
	// parse cache file into a JSON object
	jsonerr := json.Unmarshal(fileContent, &cacheFileJSON)
	if jsonerr != nil {
		fmt.Println("failed to read cache file")
		return false
	}

	// get the current date
	timenow := time.Now()

	// format date to YYYY-MM-DD
	date := timenow.Format("2006-01-02")

	// does the cache file date match the current date?
	return cacheFileJSON.Date == date
}

func main() {
	// check if the cache is outdated
	if cacheIsUpdated() {
		// cache is up-to-date, we can use it

		// fmt.Println("### yo mistah white we got cache yo")

		readWord()
	} else {
		// cache is outdated we gotta update it

		// fmt.Println("### jesse we need to fetch")

		fetchWord()
		readWord()
	}
}
